File: 백준/레벨Gold V/1717. 집합의 표현/집합의 표현.cc
#include <bits/stdc++.h>
using namespace std;

void disjointSet() {
    // 부모 노드 초기화
    int N, M;
    cin >> N >> M;

    vector<int> parent(N + 1);
    iota(parent.begin(), parent.end(), 0);

    auto find = [&](int x) -> int {
        // 경로 압축
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    };

    auto unionSets = [&](int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA > rootB) {
            parent[rootA] = rootB;
        } else {
            parent[rootB] = rootA;
        }
    };

    // 서로소 집합 연산
    while (M--) {
        int op, a, b;
        cin >> op >> a >> b;

        if (op == 0) {
            unionSets(a, b);
        } else {
            cout << (find(a) == find(b) ? "YES" : "NO") << "\n";
        }
    }
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    disjointSet();
}

File: 백준/레벨Gold V/1717. 집합의 표현/집합의 표현.java
import java.io.*;

class Solution {
    private int[] parent;

    public void disjointSet() throws IOException {
        // 부모 노드 초기화
        int N = nextInt(), M = nextInt();

        parent = new int[N + 1];
        for (int i = 0; i <= N; i++) {
            parent[i] = i;
        }

        // 서로소 집합 연산
        StringBuilder sb = new StringBuilder();
        while (M-- > 0) {
            int op = nextInt(), a = nextInt(), b = nextInt();

            if (op == 0) {
                union(a, b);
            } else {
                sb.append(find(a) == find(b) ? "YES" : "NO").append("\n");
            }
        }

        System.out.print(sb);
    }

    private int find(int x) {
        // 경로 압축
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private void union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA > rootB) {
            parent[rootA] = rootB;
        } else {
            parent[rootB] = rootA;
        }
    }

    int nextInt() throws IOException {
        int n = 0;
        int c;

        while ((c = System.in.read()) <= 32); // 탭 2, 개행 10, 공백 32

        do {
            n = n * 10 + (c - '0'); // 한자리씩 추가
        } while ((c = System.in.read()) > 32); // 구분자 : 공백

        return n;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        new Solution().disjointSet();
    }
}